Munge filenames before putting

kkellerlbl opened the pull request:
The blobstore expects a node at (for example) 5addfec2-1196-427f-a346-a12168b48a2e to be in S3 at 5a/dd/fe/5addfec2-1196-427f-a346-a12168b48a2e , but in Shock it is stored at 5a/dd/fe/5addfec2-1196-427f-a346-a12168b48a2e/5addfec2-1196-427f-a346-a12168b48a2e5addfec2-1196-427f-a346-a12168b48a2e.data .  So we have to munge the S3 destination before putting it.

In addition, the MinIO Go API changed since we wrote bulkS3upload, including (apparently) no longer returning the number of bytes uploaded, so we still need a way of reporting that (probably can just get from the on-disk file and assume that bytes transferred is the same).

kkellerlbl committed:
Update bulkS3upload.go

Fix upload path (very crudely)

kkellerlbl committed:
Update bulkS3upload.go

kkellerlbl committed:
Update bulkS3upload.go

kkellerlbl committed:
Update bulkS3upload.go

Try to use new API for minio.New()

kkellerlbl committed:
Update bulkS3upload.go

kkellerlbl committed:
Update bulkS3upload.go

kkellerlbl committed:
Update bulkS3upload.go

Minio calls need a context object

kkellerlbl committed:
Update bulkS3upload.go

FPutObject returns an UploadInfo object instead of number of bytes (and UploadInfo doesn't have bytes uploaded either :( )

kkellerlbl committed:
Update bulkS3upload.go

kkellerlbl committed:
Update bulkS3upload.go

kkellerlbl commented on bulkS3upload.go:
```diff
@@ -69,20 +69,27 @@ var errorLines = 0
 // and then waits on a channel for file paths that should be copied into the endpoint/bucket
 func copyWorker(bucket string, url string, accessID string, secretKey string, ssl bool, files <-chan string, nodeStats chan<- CopyResult, wg *sync.WaitGroup) {
 
+	ctx := context.Background()
```
The new MinIO call requires a context (and this is just the one example code used).

kkellerlbl commented on bulkS3upload.go:
```diff
@@ -69,20 +69,27 @@ var errorLines = 0
 	count := 0
 	for filePath := range files {
+		stringArray := strings.Split(filePath,"/")
+		objectPath := stringArray[0] + "/" + stringArray[1] + "/" + stringArray[2] + "/" + stringArray[3]
```
Munge the destination path to what blobstore expects.

kkellerlbl commented on bulkS3upload.go:
```diff
@@ -69,20 +69,27 @@ var errorLines = 0
+		objectPath := stringArray[0] + "/" + stringArray[1] + "/" + stringArray[2] + "/" + stringArray[3]
 		fullPath := rootDir + filePath
-		bytes, err := minioClient.FPutObject(bucket, filePath, fullPath, minio.PutObjectOptions{})
+		uploadInfo, err := minioClient.FPutObject(ctx, bucket, objectPath, fullPath, minio.PutObjectOptions{})
```
uploadInfo is an object with ETag and VersionID, which are sort of useless to us in this context.

sychan replied:
Does the compiler complain about an unused variable? You can use "_" instead of uploadinfo if it generates a compile error.

kkellerlbl replied:
The compiler did complain, yes.  I've assigned to _ instead, it's cleaner than pretending to use the variable.

kkellerlbl commented on bulkS3upload.go:
```diff
@@ -69,20 +69,27 @@ var errorLines = 0
 		}
-		nodeStats <- CopyResult{path: filePath, bytes: bytes, err: err}
+		log.Printf("ETag: %s VersionID: %s", uploadInfo.ETag,uploadInfo.VersionID)
+		nodeStats <- CopyResult{path: filePath, bytes: 0, err: err}
```
to do: return an accurate value for bytes

kkellerlbl replied:
Got from os.Stat().

kkellerlbl commented:
Here's the docs for the new MinIO FPutObject call: https://docs.min.io/docs/golang-client-api-reference#FPutObject .  I don't see a way to get bytes transferred from that.

kkellerlbl committed:
Update bulkS3upload.go

Remove log statement for each file, add ETag to nodeStats (just to use uploadInfo to avoid build error)

kkellerlbl committed:
Update bulkS3upload.go

Use local file size instead of bytes transferred

sychan reviewed (commented):
If the compiler doesn't complain about the unused return value then you can merge it, if it does complain I'd like to use "_" to discard the return value.

kkellerlbl committed:
Update bulkS3upload.go

Discard part of return value of FPutObject (don't need the etag/versionid for anything).

kkellerlbl committed:
Update bulkS3upload.go

kkellerlbl committed:
Update bulkS3upload.go

kkellerlbl committed:
Update bulkS3upload.go
